### PR TITLE
Revised IT to detect backward incompatible change 

### PR DESCRIPTION
**Pull request by findingrish**
Added a new revised IT group `BackwardCompatibilityMain`. The idea is to catch potential backward compatibility issues that may arise during rolling upgrade. 

This test group runs a docker-compose cluster with Overlord & Coordinator service on the previous druid version. 

Following env vars are required in the GHA file `.github/workflows/unit-and-integration-tests-unified.yml` to run this test
```
DRUID_PREVIOUS_VERSION -> Previous druid version to test backward incompatibility.
DRUID_PREVIOUS_VERSION_DOWNLOAD_URL -> URL to fetch the tar. 
```

Note, that this test group will remain disabled on Apache until we find a way to host druid tar in a separate location. 

#### Limitations 
* `druid-testing-tools` jar is not published. The image for the previous version still uses the 
extension from the current build. This could break in case of incompatible changes in this extension. 
In such a scenario the test should be disabled. However, this extension is primarily used to
test specific error scenarios and launch custom node role service (used in HighAvailability test group).

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: .github/workflows/reusable-revised-its.yml [hunk 1]
```diff
@@ -57,6 +57,19 @@ on:
...
+        required: false
+        type: string
+        default: false
+      DRUID_PREVIOUS_VERSION:
+        required: false
+        type: string
+      DRUID_PREVIOUS_VERSION_DOWNLOAD_URL:
+        required: false
+        type: string
+      DRUID_PREVIOUS_IT_IMAGE_NAME:
```
Comment (cryptoe) [inline .github/workflows/reusable-revised-its.yml:70]: I think we can remove this image name. We can easily derive it from the `DRUID_PREVIOUS_VERSION`

File: .github/workflows/reusable-revised-its.yml [hunk 2]
```diff
@@ -57,6 +57,19 @@ on:
       AWS_SECRET_ACCESS_KEY:
         required: false
         type: string
+      BACKWARD_INCOMPATIBILITY_IT_ENABLED:
```
Comment (cryptoe) [inline .github/workflows/reusable-revised-its.yml:60]: I think we can remove this as well. If druid previous version is null, we can just skip running the backward compatibility it.

File: integration-tests-ex/cases/cluster/Common/dependencies.yaml [hunk 3]
```diff
@@ -26,7 +26,7 @@ services:
   # See https://hub.docker.com/_/zookeeper
   zookeeper:
     # Uncomment the following when running on Apple Silicon processors:
-    # platform: linux/x86_64
+    platform: linux/x86_64
```
Comment (cryptoe) [inline integration-tests-ex/cases/cluster/Common/dependencies.yaml:29]: why is this required?

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/categories/BackwardIncompatibility.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.testsEx.categories;
+
+public class BackwardIncompatibility
```
Comment (cryptoe) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/categories/BackwardIncompatibility.java:22]:
```suggestion
public class BackwardCompatibility
```

Nit

File: integration-tests-ex/image/docker/Dockerfile [hunk 5]
```diff
@@ -46,13 +46,15 @@ ARG MARIADB_VERSION
...
+ARG DRUID_TESTING_TOOLS_VERSION
+ENV DRUID_TESTING_TOOLS_VERSION=$DRUID_TESTING_TOOLS_VERSION
 
 ENV DRUID_HOME=/usr/local/druid
 
 # Populate build artifacts
 
 COPY apache-druid-${DRUID_VERSION}-bin.tar.gz /usr/local/
-COPY druid-it-tools-${DRUID_VERSION}.jar /tmp/druid/extensions/druid-it-tools/
+COPY druid-it-tools-${DRUID_TESTING_TOOLS_VERSION}.jar /tmp/druid/extensions/druid-it-tools/
```
Comment (cryptoe) [inline integration-tests-ex/image/docker/Dockerfile:57]: Can we make loading this testing tool version optional ?

Comment (findingrish) [reply integration-tests-ex/image/docker/Dockerfile]: This doesn't seem trivial. 
The old image wouldn't run since the extension is added in the common extension list here https://github.com/apache/druid/blob/master/integration-tests-ex/cases/cluster/Common/environment-configs/common.env#L53. 
The IT framework allows extending the config file to add more extensions but there is no way to remove an extension.

File: .github/workflows/reusable-revised-its.yml (hunk 1 above)
Comment (findingrish) [reply .github/workflows/reusable-revised-its.yml:70]: Done, added a job to derive it internally.

File: integration-tests-ex/cases/cluster/Common/dependencies.yaml (hunk 3 above)
Comment (findingrish) [reply integration-tests-ex/cases/cluster/Common/dependencies.yaml]: Test changes, removed.

File: integration-tests-ex/image/docker/Dockerfile (hunk 5 above)
Comment (findingrish) [reply integration-tests-ex/image/docker/Dockerfile:56]: Adding a note on this limitation in description and test readme file.

**Review by cryptoe — APPROVED** [review]

**Comment by findingrish** [comment]
All the revised ITs have run successfully (https://github.com/apache/druid/pull/16779/commits/cc3117571768ac9db9681a7d8175a0dcdeb7e59f). 

Disabling this IT in Apache for now.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/categories/BackwardCompatibilityMain.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.testsEx.categories;
+
+public class BackwardCompatibilityMain
```
Comment (kfaraz) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/categories/BackwardCompatibilityMain.java:22]: Please add a small javadoc here, in particular pointing out what the "Main" stands for.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/MultiStageQuery.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+import org.apache.druid.testing.utils.MsqTestQueryHelper;
+import org.junit.Assert;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class MultiStageQuery
```
Comment (kfaraz) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/msq/MultiStageQuery.java:45]: Please rename to `MultiStageQueryTest`.

File: integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/BackwardCompatibilityMain/ITBCMainIndexerTest.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.apache.druid.testsEx.categories.BackwardCompatibilityMain;
+import org.apache.druid.testsEx.config.DruidTestRunner;
+import org.apache.druid.testsEx.indexer.IndexerTest;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+@RunWith(DruidTestRunner.class)
+@Category(BackwardCompatibilityMain.class)
+public class ITBCMainIndexerTest extends IndexerTest
```
Comment (kfaraz) [inline integration-tests-ex/cases/src/test/java/org/apache/druid/testsEx/BackwardCompatibilityMain/ITBCMainIndexerTest.java:30]: Can we drop the `BCMain` in the name?
It is redundant since the class is already in the backward compat package.

**Review by kfaraz — COMMENTED** [review]
@findingrish , left some minor suggestions.
